### PR TITLE
KAA-677 Compilation guide for C++ SDK (Intel Edison)

### PR DESCRIPTION
**Pull request by Dubland**
/cc @forGGe

### CODE REVIEW

File: doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md [hunk 1]
```diff
@@ -4,5 +4,79 @@ title: Intel Edison
 permalink: /:path/
 sort_idx: 60
 ---
+{% include variables.md %]
```
Comment (forGGe) [inline doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md]: `%]` -> `%}`
Also, check with `test-gh-pages.sh` script.

File: doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md [hunk 2]
```diff
@@ -4,5 +4,79 @@ title: Intel Edison
...
 
-SDK integration instructions for select platforms
+This guide explains how to build applications for Intel Edison based on the Kaa C++ endpoint SDK (further, the C++ SDK).
```
Comment (forGGe) [inline doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md]: - `This guide` -> `The guide`
- Omit `(further, the C++ SDK)`

File: doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md [hunk 3]
```diff
@@ -4,5 +4,79 @@ title: Intel Edison
...
+
+**Note:** The instructions below are for the official Intel Edison firmware.
+
+## Configuring Intel Edison board
+
+If this is the first time you use the Intel Edison technology, you have to start with configuring your board. For this purpose, refer to the official Edison getting started guide.
```
Comment (forGGe) [inline doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md]: Use separate line for every sentence

File: doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md [hunk 4]
```diff
@@ -4,5 +4,79 @@ title: Intel Edison
...
+After finishing with the guide, you will be able to access your board through both a serial terminal and an ssh client.
+
+## Installing third-party components for C++ SDK
+
+The following third-party components must be installed on the Intel Edison board before building the C++ SDK.
+
+* **Mandatory:** [Boost](http://www.boost.org/) (1.54 min.v.), [Avro](http://avro.apache.org/) (1.7.5 - 1.7.7 v.) and [Botan](https://botan.randombit.net/)  (only v.1.10) libraries.
```
Comment (forGGe) [inline doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md]: Do a list of the dependencies that consistent with other guides. See Raspberry or Artik 5 guides for examples.

File: doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md [hunk 5]
```diff
@@ -4,5 +4,79 @@ title: Intel Edison
...
+* **Optional:** [SQLite](https://www.sqlite.org/) library. It should be installed only if you are going to use a persistent log storage for [the Kaa data collection feature](https://docs.kaaproject.org/display/KAA/Collecting+data+from+endpoints).
```
Comment (forGGe) [inline doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md]: `https://docs.kaaproject.org/display/KAA/Collecting+data+from+endpoints` is old link. Use new docs link.

File: doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md [hunk 6]
```diff
@@ -4,5 +4,79 @@ title: Intel Edison
...
+
+**NOTE:** The further instructions are expected to be executed **in the order given**, directly **on the Edison board**, and **with the root permissions**.
```
Comment (forGGe) [inline doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md]: Not all these commands are required to run with root permissions. Make note about it.

File: doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md [hunk 7]
```diff
@@ -4,5 +4,79 @@ title: Intel Edison
...
+
+
+To install these libraries, proceed as follows:
+
+1. Connect to the board either though the serial terminal or an ssh client. Refer to [the official Edison getting started guide](https://software.intel.com/en-us/intel-edison-board-getting-started-guide) for any help required.
+
+2. Install the "install" utility.
```
Comment (forGGe) [inline doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md]: `coreutils` != `install`. Fix the typo.

File: doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md [hunk 8]
```diff
@@ -4,5 +4,79 @@ title: Intel Edison
...
+   ```
+    opkg install http://repo.opkg.net/edison/repo/core2-32/coreutils_8.22-r0_core2-32.ipk
+   ```
+
+3. Install Boost (1.54 or higher).
```
Comment (forGGe) [inline doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md:46]: Indent code block properly.

File: doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md [hunk 9]
```diff
@@ -4,5 +4,79 @@ title: Intel Edison
...
+    make install
+    ```
+## Creating applications based on C++ SDK
+
+Creating application based on C++ SDK for Linux platform identical for Intel Edison platform.
+You can follow [this instructions]({{root_url}Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Linux#quick-way-to-build-sdk).
+
+## Known Issues
+
+On some Edison boards, the build crashes due to the lack of system resources. If you experience this issue, reduce the number of workers used by the make utility by replacing 'make -j4' with 'make' in the build.sh script and re-runing it:
```
Comment (forGGe) [inline doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md]: For purposes of these guides we not use `build.sh`. Leave note about system resources, but drop notions about `build.sh` script.

**Comment by forGGe** [comment]
reviewed

File: doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md [hunk 10]
```diff
@@ -4,5 +4,79 @@ title: Intel Edison
...
+```
```
Comment (balldir) [inline doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md]: Please, add language tip.

> ```sh

**Comment by forGGe** [comment]
Please, squash commits into one, since there is nothing important and atomic in added changes in documentation.

File: doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md [hunk 11]
```diff
@@ -4,5 +4,93 @@ title: Intel Edison
...
+
+**NOTE:** The further instructions are expected to be executed **in the order given**, directly **on the Edison board**, and some instructions **with the root permissions**.
+
+
+To install these libraries, proceed as follows:
+
+1. Connect to the board either though the serial terminal or an ssh client.
+Refer to [the official Edison getting started guide](https://software.intel.com/en-us/intel-edison-board-getting-started-guide) for any help required.
+
+2. Install the "coreutils" utility.
```
Comment (forGGe) [inline doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md]: Emphasize "coreutils" with backticks -> `coreutils`

File: doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md [hunk 12]
```diff
@@ -4,5 +4,93 @@ title: Intel Edison
...
+- [Avro](http://avro.apache.org/) (1.7.5 - 1.7.7 v.)
+- [Botan](https://botan.randombit.net/) (only v.1.10) libraries.
+
+**Optional:**
+  
+- [SQLite](https://www.sqlite.org/) library. It should be installed only if you are going to use a persistent log storage for [the Kaa data collection feature]({{root_url}}Programming-guide/Key-platform-features/Data-collection).
+
+**NOTE:** The further instructions are expected to be executed **in the order given**, directly **on the Edison board**, and some instructions **with the root permissions**.
+
+
```
Comment (forGGe) [inline doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md]: Redundant newline

File: doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md [hunk 13]
```diff
@@ -4,5 +4,93 @@ title: Intel Edison
...
+   tar -zxf sqlite-autoconf-3081002.tar.gz
+   cd sqlite-autoconf-3081002/
+   ./configure
+   make install
+   ```
+   
+## Creating applications based on C++ SDK
+
+Creating application based on C++ SDK for Linux platform identical for Intel Edison platform.
+You can follow [this instructions]({{root_url}}Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Linux#quick-way-to-build-sdk).
```
Comment (forGGe) [inline doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md]: `You can follow [this instructions]` -> `You can follow [the Linux guide]`

File: doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md [hunk 14]
```diff
@@ -4,5 +4,93 @@ title: Intel Edison
...
+
+## Known Issues
+
+On some Edison boards, the build crashes due to the lack of system resources (out of memory).
+If you experience this issue, reduce the number of workers used by the make utility.
```
Comment (forGGe) [inline doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md]: make -> `make`

File: doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md (hunk 11 above)
Comment (rasendubi) [inline doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md]: I wouldn't emphasize at all. We don't put `Boost`, `Botan`, `Kaa` anywhere.

Comment (rasendubi) [inline doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md]: So just

```
2. Install coreutils.
```

File: doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md (hunk 14 above)
Comment (rasendubi) [inline doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md]: Or just GNU Make or Make

**Comment by Dubland** [comment]
FIxed

**Comment by forGGe** [comment]
+

File: doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md [hunk 15]
```diff
@@ -4,5 +4,92 @@ title: Intel Edison
...
+For this purpose, refer to the official Edison getting started guide.
+After finishing with the guide, you will be able to access your board through both a serial terminal and an ssh client.
+
+## Installing third-party components for C++ SDK
+
+The following third-party components must be installed on the Intel Edison board before building the C++ SDK.
+
+**Mandatory:** 
+ 
+- [Boost](http://www.boost.org/) (1.54 min.v.)
```
Comment (rasendubi) [inline doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md]: what's `min.v.`?

File: doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md [hunk 16]
```diff
@@ -4,5 +4,92 @@ title: Intel Edison
...
+- [Avro](http://avro.apache.org/) (1.7.5 - 1.7.7 v.)
```
Comment (rasendubi) [inline doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md]: `v.` should go in front if at all

File: doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md [hunk 17]
```diff
@@ -4,5 +4,92 @@ title: Intel Edison
...
+- [Botan](https://botan.randombit.net/) (only v.1.10) libraries.
```
Comment (rasendubi) [inline doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md]: We're using Botan 1.11. Botan 1.10 _does not_ work.

File: doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md [hunk 18]
```diff
@@ -4,5 +4,92 @@ title: Intel Edison
...
+   ```
+   wget https://github.com/randombit/botan/archive/1.11.28.tar.gz
+   tar -zxf 1.11.28.tar.gz
+   cd botan-1.11.28/
+   ./configure.py
+   make install
+   ln -s /usr/local/include/botan-1.11/botan /usr/local/include/botan
+   ```
+
+6. Install SQLite (latest).
```
Comment (rasendubi) [inline doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md]: I don't think 2015 is latest.

File: doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md [hunk 19]
```diff
@@ -4,5 +4,92 @@ title: Intel Edison
...
+   
+## Creating applications based on C++ SDK
+
+Creating application based on C++ SDK for Linux platform identical for Intel Edison platform.
+You can follow [the Linux guide]({{root_url}}Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Linux#quick-way-to-build-sdk).
+
+## Known Issues
+
+On some Edison boards, the build crashes due to the lack of system resources (out of memory).
+If you experience this issue, reduce the number of workers used by the Make.
```
Comment (rasendubi) [inline doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md]: How?

File: doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md [hunk 20]
```diff
@@ -4,5 +4,93 @@ title: Intel Edison
...
+
+   ```
+   wget http://sourceforge.net/projects/boost/files/boost/1.58.0/boost_1_58_0.tar.gz
+   tar -zxf boost_1_58_0.tar.gz
+   cd boost_1_58_0/
+   ./bootstrap.sh
+   ./b2 install
+   ```
+
+4. Install Avro (1.7.5-1.7.7).
```
Comment (rasendubi) [inline doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md]: Use en-dash.

File: doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md [hunk 21]
```diff
@@ -4,5 +4,93 @@ title: Intel Edison
...
+## Creating applications based on C++ SDK
+
+Creating application based on C++ SDK for Linux platform identical for Intel Edison platform.
+You can follow [the Linux guide]({{root_url}}Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Linux#quick-way-to-build-sdk).
+
+## Known Issues
+
+On some Edison boards, the build crashes due to the lack of system resources (out of memory).
+If you experience this issue, reduce the number of workers used by the Make.  
+As example: replace the **make -j4** for **make**.
```
Comment (rasendubi) [inline doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md]: Should be in backticks, not bold.

**Comment by Dubland** [comment]
Fixed.

File: doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md [hunk 22]
```diff
@@ -4,5 +4,93 @@ title: Intel Edison
...
+The guide explains how to build applications for Intel Edison based on the Kaa C++ endpoint SDK.
+
+**Note:** The instructions below are for the official Intel Edison firmware.
+
+## Configuring Intel Edison board
+
+If this is the first time you use the Intel Edison technology, you have to start with configuring your board.
+For this purpose, refer to the official Edison getting started guide.
+After finishing with the guide, you will be able to access your board through both a serial terminal and an ssh client.
+
```
Comment (vadimol) [inline doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md:18]: both the serial terminal and the ssh client

File: doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md [hunk 23]
```diff
@@ -4,5 +4,93 @@ title: Intel Edison
...
+**Optional:**
+
+- [SQLite](https://www.sqlite.org/) library. It should be installed only if you are going to use a persistent log storage for [the Kaa data collection feature]({{root_url}}Programming-guide/Key-platform-features/Data-collection).
+
+**NOTE:** The further instructions are expected to be executed **in the order given**, directly **on the Edison board**, and some instructions **with the root permissions**.
+
+To install these libraries, proceed as follows:
+
+1. Connect to the board either though the serial terminal or an ssh client.
+Refer to [the official Edison getting started guide](https://software.intel.com/en-us/intel-edison-board-getting-started-guide) for any help required.
```
Comment (vadimol) [inline doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md:38]: the ssh client

File: doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md [hunk 24]
```diff
@@ -4,5 +4,93 @@ title: Intel Edison
...
+## Creating applications based on C++ SDK
+
+Creating application based on C++ SDK for Linux platform identical for Intel Edison platform.
+You can follow [the Linux guide]({{root_url}}Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Linux#quick-way-to-build-sdk).
+
+## Known Issues
+
+On some Edison boards, the build crashes due to the lack of system resources (out of memory).
+If you experience this issue, reduce the number of workers used by the Make.  
+As example: replace the `make -j4` for `make`.
```
Comment (vadimol) [inline doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md:96]: Make -> `make`

File: doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md [hunk 25]
```diff
@@ -4,5 +4,93 @@ title: Intel Edison
...
+   tar -zxf sqlite-autoconf-3081002.tar.gz
+   cd sqlite-autoconf-3081002/
+   ./configure
+   make install
+   ```
+
+## Creating applications based on C++ SDK
+
+Creating application based on C++ SDK for Linux platform identical for Intel Edison platform.
+You can follow [the Linux guide]({{root_url}}Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Linux#quick-way-to-build-sdk).
```
Comment (vadimol) [inline doc/Programming-guide/Using-Kaa-endpoint-SDKs/C++/SDK-Edison/index.md:90]: Creating an application based on C++ SDK for Linux platform is identical for Intel Edison platform.

**Comment by vadimol** [comment]
reviewed, few grammar notes

**Comment by Dubland** [comment]
Fixed

**Comment by vadimol** [comment]
Looks good for me

**Pull request merged**